tests/commands/test_store_client.py

```diff
@@ -0,0 +1,321 @@
+    weird_filepath = tmp_path / 'not_created_dir' / 'deep' / 'credentials'
+    auth_holder._cookiejar_filepath = str(weird_filepath)
+    auth_holder._load_credentials()
+
+    # set a cookie and ask for saving it
+    auth_holder._cookiejar.set_cookie(get_cookie(value='different'))
+    auth_holder._save_credentials()
+
+    # file should be there
+    assert weird_filepath.exists()
+
+
+def test_authholder_request_simple(auth_holder):
+    """Load credentials the first time, hit the network, save credentials."""
+    # save a cookie to be used
+    fake_cookiejar = MozillaCookieJar(auth_holder._cookiejar_filepath)
+    fake_cookiejar.set_cookie(get_cookie())
+    fake_cookiejar.save()
+
+    other_cookie = get_cookie(value='different')
+
+    def fake_request(self, method, url):
+        # check it was properly called
+        assert method == 'testmethod'
+        assert url == 'testurl'
+
+        # check credentials were loaded at this time
+        assert auth_holder._cookiejar is not None
+
+        # modify the credentials, to simulate that a re-auth happened while the request
+        auth_holder._cookiejar.set_cookie(other_cookie)
+
+        return 'raw request response'
+
+    with patch('macaroonbakery.httpbakery.Client.request', fake_request):
+        resp = auth_holder.request('testmethod', 'testurl')
+
+    # verify response (the calling checks were done above in fake_request helper)
+    assert resp == 'raw request response'
+
+    # check the credentials were saved (that were properly loaded was also check in above's helper)
+    new_cookiejar = MozillaCookieJar(auth_holder._cookiejar_filepath)
+    new_cookiejar.load()
+    assert list(new_cookiejar)[0].value == other_cookie.value
+
+
+def test_authholder_request_credentials_already_loaded(auth_holder):
+    """Do not load credentials if already there."""
+    # save a cookie to be used
+    fake_cookiejar = MozillaCookieJar(auth_holder._cookiejar_filepath)
+    fake_cookiejar.set_cookie(get_cookie())
+    fake_cookiejar.save()
+
+    # load credentials, and ensure that this will be fail if it's called again
+    auth_holder._load_credentials()
+    auth_holder._load_credentials = None
+
+    with patch('macaroonbakery.httpbakery.Client.request'):
+        auth_holder.request('testmethod', 'testurl')
+
+
+def test_authholder_request_interaction_error(auth_holder):
+    """Support authentication failure while doing the request."""
+    # save a cookie to be used
+    fake_cookiejar = MozillaCookieJar(auth_holder._cookiejar_filepath)
+    fake_cookiejar.set_cookie(get_cookie())
+    fake_cookiejar.save()
+
+    with patch('macaroonbakery.httpbakery.Client.request') as mock:
+        mock.side_effect = httpbakery.InteractionError('bad auth!!')
+        expected = "Authentication failure: cannot start interactive session: bad auth!!"
+        with pytest.raises(CommandError, match=expected):
+            auth_holder.request('testmethod', 'testurl')
```

Comment:
should we be testing that this doesn't have a side effect of breaking/deleting the existing cookie file?

Reply:
We don't care, really... if the auth failed, which makes the auth processes to be attempted, which failed, the original file was already of no use.